### PR TITLE
k8s-workload-registrar: Use SPIRE generated Parent IDs

### PR DESCRIPTION
**Pull request by faisal-memon**
Signed-off-by: Faisal Memon <f.memon@f5.com>

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [ ] Documentation updated?

**Affected functionality**
CRD mode k8s-workload-registrar

**Description of change**
Currently the k8s-workload-registrar creates its own set of parent entries, one per node, using a node controller. Auto generated entries are then parented to the entry corresponding to the node the pod is running on. I realized recently that it isnt necessary to create parent entries because they are already created as part of the PSAT process.

This changeset switches to using the PSAT parents and starts the process of deprecating the node controller. To maintain backwards compatibility:
- Entries using the old node controller parent will be updated in place to use the SPIRE generated PSAT parent
- The node controller will stick around to delete the existing parents it created in an older version. In a future version this should be removed.

### CODE REVIEW

File: support/k8s/k8s-workload-registrar/mode-crd/controllers/pod_controller.go [hunk 1]
```diff
@@ -143,9 +148,10 @@ func (r *PodReconciler) updateorCreatePodEntry(ctx context.Context, pod *corev1.
 		return ctrl.Result{Requeue: true}, nil
 	}
 
-	// Check if label or annotation has changed
-	if spiffeID.Spec.SpiffeId != existing.Spec.SpiffeId {
+	// Check if label/annotation or parentId has changed
+	if spiffeID.Spec.SpiffeId != existing.Spec.SpiffeId || spiffeID.Spec.ParentId != existing.Spec.ParentId {
```
Comment (faisal-memon) [inline support/k8s/k8s-workload-registrar/mode-crd/controllers/pod_controller.go:152]: This check is what ensure the existing entries are updated

**Comment by azdagron** [comment]
Thanks for this @faisal-memon. I took a pass at it and it seems ok. I do have a few questions:
1 - How long will it take for this new controller to fix things up for pods on a given node. If the node controller runs first and removes the node registration entry, there is a chance the agent sees that before the workload registration entry is fixed up, which could end up closing down the stream with the workload API. I'm wondering if there is anything we can do to mitigate this or somehow defer reconciliation of the node stuff somehow until after the entries have been updated.
2 - I believe we support multiple registrars being deployed today (in the absence of leader election). In the presence of a new registrar and an old registrar, it seems that they will be fighting back and forth. Anything we can do to mitigate/avoid this?

**Comment by faisal-memon** [comment]
> Thanks for this @faisal-memon. I took a pass at it and it seems ok. I do have a few questions:
> 1 - How long will it take for this new controller to fix things up for pods on a given node. If the node controller runs first and removes the node registration entry, there is a chance the agent sees that before the workload registration entry is fixed up, which could end up closing down the stream with the workload API. I'm wondering if there is anything we can do to mitigate this or somehow defer reconciliation of the node stuff somehow until after the entries have been updated.

When the new registrar starts it gets reconcile events for everything. So everything will start getting updated right away, how long it takes depends on how many entries there. Is there a way to know how many children a parent entry has. We can wait till that number is 0 before deleting the node parent entry.

> 2 - I believe we support multiple registrars being deployed today (in the absence of leader election). In the presence of a new registrar and an old registrar, it seems that they will be fighting back and forth. Anything we can do to mitigate/avoid this?

Thats a tough one because the two arent compatible to run side by side. let me think about that case.

**Comment by azdagron** [comment]
> When the new registrar starts it gets reconcile events for everything
Is there a way to detect that all of the initial reconcile events are done? If so,  can we just run the pod reconciler against the initial events first before starting up the node reconciler?

**Comment by faisal-memon** [comment]
> > When the new registrar starts it gets reconcile events for everything
> 
> Is there a way to detect that all of the initial reconcile events are done? If so, can we just run the pod reconciler against the initial events first before starting up the node reconciler?

unfortunately i don't know of a good way to do that. it works best when these things are self-contained.

**Comment by faisal-memon** [comment]
> > > When the new registrar starts it gets reconcile events for everything
> > 
> > 
> > Is there a way to detect that all of the initial reconcile events are done? If so, can we just run the pod reconciler against the initial events first before starting up the node reconciler?
> 
> unfortunately i don't know of a good way to do that. it works best when these things are self-contained.

I put in a commit that has the node reconciler wait until there are no children before deleting the parent.

**Comment by faisal-memon** [comment]
>2 - I believe we support multiple registrars being deployed today (in the absence of leader election). In the presence of a new registrar and an old registrar, it seems that they will be fighting back and forth. Anything we can do to mitigate/avoid this?

Did a lot of testing for this scenario. 
- If the two different versions are running side by side but nothing is deployed eventually everything reconciles ok when the old version is removed, except for https://github.com/spiffe/spire/issues/2275 i reported earlier, where a duplicate psat entry will be created. I can push a commit that will clean up this duplicate entry if thats helpful.
- Deploying stuff when the two different versions are running does not work well. Having multiple k8s-workload-registrar's works ok when they are trying to create the same entries. When they are trying to create different ones it quickly gets out of sync.
In the end I think the safest way to go about upgrading is to first set `pod_controller = false` which disable new entries from being auto generated and won't run the pod or node reconcilers. Then upgrade and reenable `pod_controller = true`.

**Comment by evan2645** [comment]
I've been thinking about this change and how we can do it safely. I think it is likely that folks can get into a spot where they have both the "old" and the "new" version deployed simultaneously... so making sure that we can cope with this condition is really important IMO

The best answer is fully automatic detection :) but I know that may not be easy... for example, if we could somehow detect that an older version is running, and alter our behavior accordingly, that would be awesome.

If that's not possible, then the only thing I can come up with is a configurable-based approach, as @faisal-memon has suggested. I think one big challenge with this is, how do we educate users that a special thing has to be done now? And what is the consequence if they upgrade without seeing the message or doing the necessary thing?

Finally, I try to weigh this against the gain. In this case, we get simplified and fewer entries... is there anything else we get?

> I think one big challenge with this is, how do we educate users that a special thing has to be done now? And what is the consequence if they upgrade without seeing the message or doing the necessary thing?

Since (I think?) we're talking about a small optimization here that many folks won't notice or see, I think we need really good answers to these questions. For example, I think the cost/benefit doesn't work out so well if the answer to the second question is that an outage is possible.

**Comment by faisal-memon** [comment]
Shelving this until we can think of a non-disruptive way of rolling out these changes.

**Pull request closed without merging**